source/routing/journey_pattern_container.cpp

```diff
@@ -0,0 +1,217 @@
+    const auto& jp = get(jp_from_vj[VjIdx(*st.vehicle_journey)]);
+    return jp.jpps.at(st.order());
+}
+
+std::string JourneyPatternContainer::get_id(const JpIdx& jp_idx) const {
+    std::stringstream ss;
+    ss << "journey_pattern:" << jp_idx.val;
+    return ss.str();
+}
+std::string JourneyPatternContainer::get_id(const JppIdx& jpp_idx) const {
+    std::stringstream ss;
+    ss << "journey_pattern_point:" << jpp_idx.val;
+    return ss.str();
+}
+boost::optional<JpIdx> JourneyPatternContainer::get_jp_from_id(const std::string& id) const {
+    static const std::string prefix = "journey_pattern:";
+    if (id.size() > prefix.size() && id.substr(0, prefix.size()) == prefix) {
+        try {
+            const auto idx = boost::lexical_cast<idx_t>(id.substr(prefix.size()));
+            if (idx < nb_jps()) { return JpIdx(idx); }
+        } catch(boost::bad_lexical_cast&) {}
+    }
+    return boost::none;
+}
+boost::optional<JppIdx> JourneyPatternContainer::get_jpp_from_id(const std::string& id) const {
+    static const std::string prefix = "journey_pattern_point:";
+    if (id.size() > prefix.size() && id.substr(0, prefix.size()) == prefix) {
+        try {
+            const auto idx = boost::lexical_cast<idx_t>(id.substr(prefix.size()));
+            if (idx < nb_jpps()) { return JppIdx(idx); }
+        } catch(boost::bad_lexical_cast&) {}
+    }
+    return boost::none;
+}
+
+template<typename VJ>
+JourneyPatternContainer::JpKey JourneyPatternContainer::make_key(const VJ& vj) {
+    JourneyPatternContainer::JpKey key;
+    key.route_idx = RouteIdx(*vj.route);
+    key.phy_mode_idx = PhyModeIdx(*vj.physical_mode);
+    key.is_freq = std::is_same<VJ, nt::FrequencyVehicleJourney>::value;
+    for (const auto& st: vj.stop_time_list) {
+        key.jpp_keys.emplace_back(SpIdx(*st.stop_point),
+                                  st.local_traffic_zone,
+                                  st.properties);
+    }
+    return key;
+}
+
+static bool st_are_equals(const nt::VehicleJourney& vj1, const nt::VehicleJourney& vj2) {
+    if (vj1.stop_time_list.size() != vj2.stop_time_list.size()) { return false; }
+    for (auto it1 = vj1.stop_time_list.begin(), it2 = vj2.stop_time_list.begin();
+         it1 !=  vj1.stop_time_list.end() && it2 !=  vj2.stop_time_list.end();
+         ++it1, ++it2) {
+        if (it1->arrival_time != it2->arrival_time || it1->departure_time != it2->departure_time) {
+            return false;
+        }
+    }
+    return true;
+}
+
+template<typename VJ> static bool
+overtake(const VJ& vj, const std::vector<const VJ*> vjs) {
+    if (vj.stop_time_list.empty()) { return false; }
+    for (const VJ* cur_vj: vjs) {
+        assert(vj.stop_time_list.size() == cur_vj->stop_time_list.size());
+
+        // if the validity patterns do not overlap, it can't overtake
+        static const auto levels = {nt::RTLevel::Theoric, nt::RTLevel::Adapted, nt::RTLevel::RealTime};
```

Comment:
we just cannot iterate over a enum, shame for C++11... Ada has this feature from 80's... :(